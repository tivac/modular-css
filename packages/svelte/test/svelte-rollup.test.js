"use strict";

const fs = require("fs");

const rollup  = require("rollup").rollup;

const read     = require("test-utils/read.js")(__dirname);
const namer    = require("test-utils/namer.js");
const watching = require("test-utils/rollup-watching.js");

const processor = require("../rollup.js");

function cleanup(code) {
    return code.replace(/\/\* packages.*$/gm, "");
}

const assetFileNames = "assets/[name][extname]";
const format = "es";
const map = false;
const sourcemap = false;

const output = "./packages/svelte/test/output";

describe("/rollup.js", () => {
    // afterEach(() => require("shelljs").rm("-rf", `${output}/*`);
    
    it("should generate exports", () => {
        const { preprocess, plugin } = processor({
            css : `${output}/svelte.css`,
            namer,
        });

        return rollup({
            input   : require.resolve("./specimens/style.html"),
            plugins : [
                require("rollup-plugin-svelte")({
                    preprocess
                }),
                plugin
            ]
        })
        .then((bundle) => bundle.write({
            format,
            assetFileNames,
            sourcemap,
            
            file : `${output}/svelte.js`
        }))
        .then(() => {
            expect(cleanup(read("svelte.js"))).toMatchSnapshot();
            expect(read("svelte.css")).toMatchSnapshot();
        });
    });

    it("should not generate exports", () => {
        const { preprocess, plugin } = processor({
            namer,
        });

        return rollup({
            input   : require.resolve("./specimens/style.html"),
            plugins : [
                require("rollup-plugin-svelte")({
                    preprocess
                }),
                plugin
            ]
        })
        .then((bundle) => bundle.write({
            format,
            assetFileNames,
            sourcemap,
            
            file : `${output}/svelte.js`
        }))
        .then(() => {
            expect(() => read("svelte.css")).toThrow();
        });
    });

    it("should support external css w/o a <script>", async () => {
        const { preprocess, plugin } = processor({
            namer,
            map,
        });

        const bundle = await rollup({
            input   : require.resolve("./specimens/external.html"),
            plugins : [
                require("rollup-plugin-svelte")({
                    preprocess
                }),
                plugin
            ]
        });

        await bundle.write({
            format,
            assetFileNames,
            sourcemap,
            
            file : `${output}/svelte.js`
            
        });

        expect(cleanup(read("svelte.js"))).toMatchSnapshot();
        expect(read("assets/svelte.css")).toMatchSnapshot();
    });

    it("should support external css w/ a script", async () => {
        const { preprocess, plugin } = processor({
            namer,
            map,
        });

        const bundle = await rollup({
            input   : require.resolve("./specimens/external-script.html"),
            plugins : [
                require("rollup-plugin-svelte")({
                    preprocess
                }),
                
                plugin
            ]
        });

        await bundle.write({
            format,
            assetFileNames,
            sourcemap,
            
            file : `${output}/svelte.js`
        });

        expect(cleanup(read("svelte.js"))).toMatchSnapshot();
        expect(read("assets/svelte.css")).toMatchSnapshot();
    });

    describe("rollup watcher", () => {
        const watch = require("rollup").watch;
        let watcher;

        afterEach(() => watcher.close());

        it("should support file changes when using watch", (done) => {
            const { preprocess, plugin } = processor({
                namer,
                map,
            });

            // Create HTML
            fs.writeFileSync(
                `${output}/watch.html`,
                `<link rel="stylesheet" href="./watched.css" />`
            );

            // Create v1 of the file
            fs.writeFileSync(
                `${output}/watched.css`,
                ".one { color: red; }"
            );

            // Start watching
            watcher = watch({
                input  : require.resolve("./output/watch.html"),
                output : {
                    file : `${output}/watch-output.js`,
                    format,
                    assetFileNames,
                },
                plugins : [
                    require("rollup-plugin-svelte")({
                        preprocess
                    }),
                    plugin
                ]
            });

            // Create v2 of the file after a bit
            setTimeout(() => fs.writeFileSync(
                "./packages/rollup/test/output/watched.css",
                ".two { color: blue; }"
            ), 200);
            
            watcher.on("event", watching((builds) => {
                console.log(builds);
                
                if(builds === 1) {
                    console.log(read("watch-output.js"));

                    expect(read("assets/watch-output.css")).toMatchSnapshot();

                    // continue watching
                    return null;
                }

                expect(read("assets/watch-output.css")).toMatchSnapshot();

                return done();
            }));
        });
    });
});
