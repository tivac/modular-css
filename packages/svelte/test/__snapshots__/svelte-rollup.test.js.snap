// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/svelte.js rollup chunking should correctly chunk svelte files using inline <style> 1`] = `
Object {
  "a-dep.css": "
/* packages/svelte/test/specimens/inline-chunking/a-dep.css */
.mc6ca450b0_adep {
    margin: 5px;
}
",
  "a.css": "
/* packages/svelte/test/specimens/inline-chunking/a.svelte */
.mc95c9e282_a {
    color: aqua;
}
",
  "b-dep.css": "
/* packages/svelte/test/specimens/inline-chunking/b-dep.css */
.mcf56d52f6_bdep {
    padding: 5px;
}
",
  "b.css": "
/* packages/svelte/test/specimens/inline-chunking/b.svelte */
.mc39b698fc_b {
    color: blue;
}
",
  "shared.css": "
/* packages/svelte/test/specimens/inline-chunking/shared.css */
.mcfef68931_shared {
    background: black;
}
",
}
`;

exports[`/svelte.js rollup errors should show useful errors from rollup (non-css file) 1`] = `
Array [
  Array [
    "Possible invalid <link> href: ./error-link.svelte",
  ],
]
`;

exports[`/svelte.js rollup watching should generate updated output 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
+   "assets/app.css": "/* packages/svelte/test/output/rollup/input/app.css */
+ .mc043115d5_nope {
+     color: blue;
+ }",
@@ --- --- @@
- 			attr(div, \\"class\\", \\"css.nope\\");
+ 			attr(div, \\"class\\", \\"mc043115d5_nope\\");
`;

exports[`/svelte.js rollup watching should generate updated output when composition changes 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@

  	return {
  		c() {
  			div = element(\\"div\\");
  			div.textContent = \\"Hi\\";
- 			attr(div, \\"class\\", \\"mc8c3b9f40_b mc7c0d2c82_a\\");
+ 			attr(div, \\"class\\", \\"mc8c3b9f40_c mc7c0d2c82_a\\");
  		},
  		m(target, anchor) {
  			insert(target, div, anchor);
  		},
  		p: noop,
`;
